### ISSUE TITLE
Returning records from cache seems to not work.

### ISSUE THREAD

**Issue by slightlytyler** [tonyhb/tectonic#57]
`We check each query to see if we have cached data already available to skip unnecessary API requests. If so, this data is passed to the component` is what I'm referencing. In my case I have a `Table` component which loads a list and renders each `TableRow` with an `id` prop. Each `TableRow` loads an item and display its content. The problem I'm experiencing is that a network request is being made for each `TableRow` rather than the record being passed from the store.

**Comment by tonyhb** [comment]
Hmm, taking a look at this now.  Just a note - the cache respects server side headers.  Could you paste the cache-control and expires headers from the API request?

We do this because we've ran into a lot of situations in which we need to either force-refresh data or get it from the store.  Seeing as browsers and APIs have the ability to do this based on these headers that's what we use.

If it's due to these headers we may need an operator to force caching for some amount of time.  If not then perhaps there's a bug

**Comment by slightlytyler** [comment]
@tonyhb don't think the API I'm working with defines either of those. Just for my knowledge are you suggesting these headers on the GET request for the list affect the caching of the GET for the item?

**Comment by tonyhb** [comment]
Yep.  It tries to stick to conventions, so if the endpoint doesn't specify that stale data is okay via caching headers we also respect that.

If that's unexpected we need a way to override these API headers with defaults in 2.1, which we can roll out today for you.

We need a nice API though, so we can:

1. Define custom expiry times by each model (ie in `class User extends Model`)
2. Define custom expiry times by each API endpoint
3. Force a custom expiry within `@load` (`user: User.getItem({ id: 1 }).cacheFor(60)` or something).

Thoughts?

**Comment by tonyhb** [comment]
I'm thinking a combination of 1 and 2, using the lowest TTL so that we pessimistically cache. Say you allow all user models to be cached for an hour, but you want one endpoint to always return new data.  If this endpoint specifies 0 in the default cache, we always force new data when this API endpoint is used.

This is going to take a little work within the resolver to solve; we have to modify models and source definitions to handle custom expiration time, plus the difference between an undefined and zero default expiration time.

**Comment by slightlytyler** [comment]
@tonyhb definitely agree with 1 & 2. A lot of times (like in my example) you don't have control over the API and need to override the cache behavior. Don't think 3 is needed but might also be useful to allow the user to load directly from cache without making a network request via `load`. Not sure what the syntax would look like for that. Or maybe expose a selector for doing that? Idk.

**Comment by tonyhb** [comment]
:+1: I'll start planning 1+2 after standup and meetings. Hopefully we'll have some test cases and a POC ready today/tomorrow, then we can roll out 2.1 after docs changes.

Thanks for the heads up, and sorry that from 2.0 it'll pull from the API!

**Comment by slightlytyler** [comment]
```js
load(props => ({
  records: {
    query: Model.getItem({ id: props.id }),
    options: {
      fromCache: true,
    },
  },
}))
```
Idk just a thought. So far really enjoying the library, looking forward to hopefully using it to replace my hand rolled system

**Comment by tonyhb** [comment]
Yeah, we'd definitely need a way to show query options for the future.  Planning to create a localStorage driver adapter so that we can cache API responses locally, allowing you to boot with "stale" data the next time a user loads the page.

In that case, when the localStorage returns successfully but the API returns an error we need to handle this special case successfully.  1+2 should be easy to do now, but 3 sounds a little more complex with future features.

Super glad to hear it's working well for you though!  Feel free to ping with any other issues :)

**Comment by tonyhb** [comment]
Whoa took a while! @slightlytyler fixed in https://github.com/tonyhb/tectonic/pull/97, will roll out 2.2.0 tonight.

**Comment by tonyhb** [comment]
Sorry for the late update.  Left the job at Docker and went to a tiny startup... so much work!

**Comment by tonyhb** [comment]
Fixed in #97
